core/domain/acl_decorators.py

```diff
@@ -1033,3 +1056,41 @@ def test_can_manage_topic_rights(self, topic_id, **kwargs):
     test_can_manage_topic_rights.__wrapped__ = True
 
     return test_can_manage_topic_rights
+
+
+def can_accept_suggestion(handler):
+    """Decorator to check if user can accept the given suggestion."""
+    def test_can_accept(self, suggestion_id, **kwargs):
+        """Checks if the user can accept the suggestion.
+
+        Args:
+            suggestion_id: str. The suggestion id.
+            **kwargs: *. Keyword arguments.
+
+        Returns:
+            *. The return value of the decorated function.
+        """
+        if not self.user_id:
+            raise base.UserFacingExceptions.NotLoggedInException
+
+        print self.user
+        if len(suggestion_id.split('.')) != 3:
+            raise self.InvalidInputException('Invalid format for suggestion_id.'
+                                             ' It must contain 3 parts'
+                                             ' separated by \'.\'')
+        target_type = suggestion_id.split('.')[0]
+
+        if target_type == suggestion_models.TARGET_TYPE_EXPLORATION:
```

Comment:
Just a thought - this could result in a lot of if..else statements. Would keeping this in the domain layer be worth it?

Reply:
I thought about this actually. So this is specific to the target type and not the suggestion type, so if we abstract to the domain layer, we might have a lot of repetition of this function (All of the exploration specific suggestions will need the same validator). So I think it is a tradeoff between repeating this much code or having if...else branches. What's your opinion on this @anmolshkl @seanlip

Reply:
These are decorators, right? I would say it depends on how you want to structure access control for this. If it's similar for all objects, then it's fine to keep the logic in one place and reuse it. If it's different then do stuff like can_accept_exploration_edit_suggestion, can_accept_topic_edit_suggestion, etc. -- or even reuse stuff like can_edit_exploration, etc. (since someone who's accepting a suggestion is basically making a direct edit to the item with the suggested content).

One thing to note is that you want these checks in the controller/decorator layer. If someone is not authorized to do something then detect that as soon as possible and don't let the flow even end up in the domain layer.

Reply:
Actually what you said sounds better! I'll try to reuse the other decorators (is it possible to call a decorator from another one? any pointers here would be great!)

Reply:
Don't know that we've done that before, sorry (you'll need to research).

But if you have separate handlers for the various types of suggestions, perhaps you could also just use e.g. can_edit_exploration, etc. directly for the relevant handler.

Reply:
Ah that sounds like a good idea! This should give us better control I think.

Reply:
I did try to use a decorator in here -- structurally it feels weird but it works! This way we can use other decorators to actually check permissions (do not need to replicate stuff).

PTAL and give your opinions (if not this approach, then I'll have separate controllers for suggestions to each target type and use the can_edit_(TARGET-TYPE) decorator appropriately).

Reply:
That is weird :) I can't say I like it, especially all the unused stuff -- it's too implicit IMO.

I think the alternative approach you suggest is fine. It might be worth a quick chat with @1995YogeshSharma to get his thoughts on how best to do this, since he implemented the decorator system and ACL checks.

Reply:
Sure! I'll consult Yogesh for his opinion too! :) yeah the unused stuff was actually odd!